cmd/config.go

```diff
@@ -92,13 +129,113 @@ func (f *Furyconf) Parse() ([]Package, error) {
 				}
 			}
 		}
-		block := strings.Split(pkgs[i].Name, "/")
-		if len(block) == 2 {
-			pkgs[i].url = fmt.Sprintf("%s-%s%s//%s/%s?ref=%s", repoPrefix, block[0], dotGitParticle, pkgs[i].kind, block[1], version)
-		} else if len(block) == 1 {
-			pkgs[i].url = fmt.Sprintf("%s-%s%s//%s?ref=%s", repoPrefix, block[0], dotGitParticle, pkgs[i].kind, version)
+		kindSpec := newKind(kind, f.Provider)
+		if registry {
+			urlPrefix = kindSpec.pickCloudProviderURL(cloud)
+			dotGitParticle = ".git"
 		}
-		pkgs[i].dir = fmt.Sprintf("%s/%s/%s", f.VendorFolderName, pkgs[i].kind, pkgs[i].Name)
+		block := strings.Split(pkgs[i].Name, "/")
+		url = newURL(urlPrefix, block, dotGitParticle, kind, version, registry)
+		pkgs[i].url = url.strategy()
+		directory = newDir(f.VendorFolderName, pkgs[i].kind, pkgs[i].Name, pkgs[i].Registry, pkgs[i].Provider)
+		pkgs[i].dir = directory.strategy()
 	}
+
 	return pkgs, nil
 }
+
+func newKind(kind string, provider ProviderPattern) ProviderKind {
+	providerChoosen := provider[kind]
+	return providerChoosen
+}
+
+func (k *ProviderKind) getLabeledURI(providerName, label string) (string, error) {
+	for name, providerSpecList := range *k {
+		if name == providerName {
+			for _, providerMap := range providerSpecList {
+				if providerMap.Label == label {
+					return fmt.Sprintf("git::%s", providerMap.BaseURI), nil
+				}
+			}
+		}
+	}
+	return "", fmt.Errorf("no label %s found", label)
+}
+
+func (k *ProviderKind) pickCloudProviderURL(cloudProvider ProviderSpec) string {
+	name := cloudProvider.Name
+	label := cloudProvider.Label
+	url, err := k.getLabeledURI(name, label)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return url
+}
+
+// DirSpec is the abstraction of the fields needed for generating a destination directory
+type DirSpec struct {
+	VendorFolder string
+	Kind         string
+	Name         string
+	Registry     bool
+	Provider     ProviderSpec
+}
+
+func newDir(folder, kind, name string, registry bool, provider ProviderSpec) *DirSpec {
+	return &DirSpec{
+		VendorFolder: folder,
+		Kind:         kind,
+		Name:         name,
+		Registry:     registry,
+		Provider:     provider,
+	}
+}
+
```

Comment:
no need for the constructor if the DirSpec structure is public and the constructor is not enforcing anything.

Reply:
you're right, i only create the constructor for a possible enhancement that could be done in the future, in this way the environment is already setup!